Submitting for review

st-murray opened the pull request.

st-murray committed:
first code changes

greg-hellings commented:
Gonna have to do some magic about these two different Travis CI builds. The repo rename has caused two jobs to be building - one in the old "wipefs" namepsace which is failing because of the rename and the new one under the newly detected name. But that shouldn't hold up reviewing.

robled commented on README.md:
```diff
@@ -21,24 +32,48 @@ Currently the following variables are supported:
+- hosts: disk_wipe-servers
+  roles:
+    - role: oasis_roles.disk_wipe
+      vars:
```
Is this indented correctly?

st-murray replied:
updated

robled commented on README.md:
```diff
@@ -1,9 +1,20 @@
 
-Basic description for wipefs
+This role is not idempotent and is intended to be used when you are preparing nodes
+for deployment to ensure there is no stale data from previous deployments.
```
Stale data from previously deployments is only one possibility.  It's possible to buy a brand new disk and have it contain partitions directly from the manufacturer.

st-murray replied:
updated

robled commented on README.md:
```diff
@@ -1,9 +1,20 @@
+1. Ensure parted is installed.
+2. Erase the the disk signatures if detected from specified disks.
+3. Re-read the partition tables for specified disks.
+4. Completely erase all traces of existing data on specified disks.
```
This would only be true if we were to `dd` or `shred` the entire disk from beginning to end.

st-murray replied:
updated

robled commented on README.md:
```diff
@@ -1,9 +1,20 @@
 ===========
 
-Basic description for wipefs
+This role is not idempotent and is intended to be used when you are preparing nodes
```
During yesterday's discussion we agreed that we should have the README recommend to use this role in a separate, isolated playbook to prevent disks from getting wiped a second time.

st-murray replied:
updated

robled commented on README.md:
````diff
@@ -21,24 +32,48 @@ Currently the following variables are supported:
+
+```
 
+This example will run against a list of disks dev/vdb, /dev/vdc, dev/vdd
````
In my opinion this is redundant.  I think that users of OASIS roles will know how lists work as a pre-requisite.

st-murray replied:
updated

robled commented on README.md:
```diff
@@ -21,24 +32,48 @@ Currently the following variables are supported:
   roles:
-    - role: oasis_roles.wipefs
+    - role: oasis_roles.disk_wipe
+      vars:
```
Is this indented correctly?

st-murray replied:
updated

robled commented on defaults/main.yml:
```diff
@@ -1,2 +1,16 @@
+# List of disks to run the role on.
+disk_wipe_disks: []
+
+# Example
```
Given that this role is so highly dangerous, I think the examples in the README are sufficient for documentation, and listing common disk paths here, even commented out, could lead to a user typo that could create data loss.  Best to leave this out in my opinion.

"List of paths to disk block devices" would serve as a sufficient comment.

st-murray replied:
updated

robled commented on README.md:
```diff
@@ -21,24 +32,48 @@ Currently the following variables are supported:
   functionality for privilege escalation, then this is the name of the target
   user to change to.
 
+* `disk_wipe_disks` - Default: [] empty list.
```
For clarification, this documentation should read that it expects a list of paths to disk block devices.

greg-hellings replied:
I would also reiterate, here, that they are the disks which are going to be wiped and that partitions are not acceptable - they must be actual block devices.

st-murray replied:
updated

robled commented on molecule/docker/molecule.yml:
```diff
@@ -7,14 +7,14 @@ lint:
-  - name: wipefs
-    image: centos:7
+  - name: disk_wipe
+    #image: centos:7
```
Why is this commented out?

st-murray replied:
fixed

robled commented on molecule/docker/molecule.yml:
```diff
@@ -7,14 +7,14 @@ lint:
-    # command: /usr/lib/systemd/systemd --system
-    # volumes:
-    #   - /sys/fs/cgroup:/sys/fs/cgroup:ro
+    image: centos/systemd
```
I'm hazarding a guess that we can't actually wipe disks with Travis CI/Docker.  In that case, is there any reason to use systemd here?

st-murray replied:
fixed

robled commented on tasks/main.yml:
```diff
@@ -1 +1,31 @@
+  become_user: "{{ disk_wipe_become_user }}"
+  command: "wipefs --all --force {{ item }}"
+  register: signatures_before
+  changed_when: '"offset" in signatures_before.stdout'
```
Given that this is running unconditionally, and we're using the `--force` flag to `wipefs`, and that this role is not idempotent, we might as well just report 'changed' status here.

greg-hellings replied:
I mean, if we can detect that something has or has not changed, we should. But are we sure that this conditional is correctly picking up whether something has changed?

st-murray replied:
updated as my intent is to run unconditionally

robled commented on tasks/main.yml:
```diff
@@ -1 +1,31 @@
+  changed_when: '"offset" in signatures_before.stdout'
+  with_items: "{{ disk_wipe_disks }}"
+
+- name: Re-read partion tables
```
This should probably run at the end, given that we are performing more disk write activities _after_ this.  Even so, is this necessary in the context of just `wipefs`?  From the `wipefs` man page:

> wipefs calls the BLKRRPART ioctl when it has erased a partition-table signature to inform the kernel about the change.

st-murray replied:
re-ordered role to run shred, wipefs no longer a need for parted to re-read partition tables.

robled commented on tasks/main.yml:
```diff
@@ -1 +1,31 @@
+  become: "{{ disk_wipe_become }}"
+  become_user: "{{ disk_wipe_become_user }}"
+  command: "partprobe {{ item }}"
+  changed_when: false
```
Technically by changing the list of partitions that Linux can see, we are making a change here.

st-murray replied:
updated

robled commented on tasks/main.yml:
```diff
@@ -1 +1,31 @@
+  become: "{{ disk_wipe_become }}"
+  become_user: "{{ disk_wipe_become_user }}"
+
+- name: Clear the disk signatures
```
I would change the name of this so that the user knows we're running the `wipefs` command.

st-murray replied:
updated

st-murray replied:
updated

robled commented on tasks/main.yml:
```diff
@@ -1 +1,31 @@
+  changed_when: false
+  with_items: "{{ disk_wipe_disks }}"
+
+- name: Erase the disk data
```
In this case we are running `shred` to write zeroes to a specific part of the disk.  The name of this task should reflect that.  

Unlike `wipefs`, I believe that partition deletion using `shred` would probably require notifying the kernel about changes to partitions.  

It might make sense to run `shred` before `wipefs`, which may allow us to drop the `parted` dependency.

st-murray replied:
Advice taken and role simplified.  run shred first and then wipefs as it will call the BLKRRPART ioctl when it has erased a partition-table signature to inform the kernel about the change.

robled commented on tasks/main.yml:
```diff
@@ -1 +1,31 @@
+- name: Erase the disk data
+  become: "{{ disk_wipe_become }}"
+  become_user: "{{ disk_wipe_become_user }}"
+  command: "shred -v -n 1 --random-source=/dev/zero -s 200M {{ item }}"
```
The **200M** value here is hard-coded.  This should probably be user-configurable, as the **200M** value was specific to a single application we were testing, and there may be others that require a different value.

st-murray replied:
updated to use variable

greg-hellings commented on README.md:
```diff
@@ -21,24 +32,48 @@ Currently the following variables are supported:
+  roles:
+    - role: oasis_roles.disk_wipe
+      vars:
+      disk_wipe_disks: /dev/vdb
```
Are we really wanting to support both a single string as well as a list?

st-murray replied:
updated

greg-hellings commented on handlers/main.yml:
```diff
@@ -1 +1 @@
-# Handlers for wipefs
+# Handlers for disk_wipe
```
If there aren't any handlers, you should get rid of this file.

st-murray replied:
removed

st-murray replied:
removed

greg-hellings commented on molecule/openstack/tests/test_null.py:
```diff
@@ -1,8 +0,0 @@
-# Without at least a file here, tests in the additional directory will not
```
These files are necessary for any shared tests to work. Do we anticipate writing tests in the future? If so, we should leave the two `test_null.py` files in place.

st-murray replied:
at this point no tests planned.

greg-hellings commented on vars/main.yml:
```diff
@@ -1 +1 @@
-# Vars for wipefs
+#Vars for disk_wipe
```
If there are no vars, please delete this file.

st-murray replied:
removed

st-murray replied:
removed

greg-hellings commented on molecule/openstack/molecule.yml:
```diff
@@ -7,7 +7,9 @@ lint:
 platforms:
-  - name: test-wipefs
+  - name: test-disk_wipe
+    image: rhel-7.4-server-x86_64-released
```
7.4 is old. Let's do this with 7.6, which is current latest, I believe.

st-murray replied:
updated

greg-hellings commented on tasks/main.yml:
```diff
@@ -1 +1,31 @@
+  become_user: "{{ disk_wipe_become_user }}"
+  command: "wipefs --all --force {{ item }}"
+  register: signatures_before
+  changed_when: true
```
This is redundant. The `command` module should always report a status of "changed" when it is executed without a `changed_when` directive or a `creates` or `removes` argument.

st-murray replied:
By default I am passing in an empty list of disks.   Without the changed when command travis/ci will not pass with the following error.
Commands should not change things if nothing needs doing

robled replied:
In this case you should flag the linter to ignore that warning.

st-murray commented:
Lot's of changes.   Simplified the role by running shred first, then wipefs to remove the need for parted.  Added variable for disk size in bytes to shred.   Plus lots of documentation changes.

robled commented on tasks/main.yml:
```diff
@@ -1 +1,17 @@
+- name: Run shred to write zeros to the disk
+  become: "{{ disk_wipe_become }}"
+  become_user: "{{ disk_wipe_become_user }}"
+  command: 'shred -v -n 1 --random-source=/dev/zero -s \
```
Use a YAML block scalar here.

st-murray replied:
Updated.  Looks much better.   Thanks

robled commented on tasks/main.yml:
```diff
@@ -1 +1,17 @@
+  become_user: "{{ disk_wipe_become_user }}"
+  command: 'shred -v -n 1 --random-source=/dev/zero -s \
+  "{{ disk_wipe_write_size }}" "{{ item }}"'
+  changed_when: true
```
`changed_when` is redundant

st-murray replied:
Added 301 to skip list and removed changed_when

robled replied:
The skip list is global.  Since we are focused on just a couple of specific `command` tasks, we should just flag those specific tasks as ignored by the linter.

st-murray replied:
OK added a tag to those to commands

robled commented on tasks/main.yml:
```diff
@@ -1 +1,17 @@
+- name: Run wipefs to clear the disk signatures
+  become: "{{ disk_wipe_become }}"
+  become_user: "{{ disk_wipe_become_user }}"
+  ignore_errors:
```
Is this syntactically correct?  Do we need to ignore errors?

st-murray replied:
Yeah my bad missed some clean up.

greg-hellings commented on molecule/openstack/molecule.yml:
```diff
@@ -7,7 +7,9 @@ lint:
-  - name: test-wipefs
+  - name: disk_wipe
+    image: rhel-7.6-server-x86_64-released
+    flavor: ci.m1.small.ephemeral
```
Why'd you go with the ephemeral instance?

st-murray commented:
Merging into master
